clients/githubrepo/branches.go

```diff
@@ -266,5 +422,166 @@ func getBranchRefFrom(data *branch) *clients.BranchRef {
+					*branchRef.BranchProtectionRule.AllowForcePushes = false
+					modded = true
+				}
+			case "PULL_REQUEST":
+				if applyPullRequestRepoRule(branchRef, rule) {
+					modded = true
+				}
+			case "REQUIRED_STATUS_CHECKS":
+				if applyRequiredStatusChecksRepoRule(branchRef, rule) {
+					modded = true
+				}
+			}
+		}
+		if modded {
+			adminEnforced := ruleAdminEnforced(r)
+			branchRef.BranchProtectionRule.EnforceAdmins = &adminEnforced
+		}
+	}
+}
+
+func applyPullRequestRepoRule(branchRef *clients.BranchRef, rule *repoRule) bool {
+	var modded bool
+	branchRules := branchRef.BranchProtectionRule.RequiredPullRequestReviews
+	dismissStale := readBoolPtr(rule.Parameters.PullRequestParameters.DismissStaleReviewsOnPush)
+	if dismissStale && !readBoolPtr(branchRules.DismissStaleReviews) {
+		branchRef.BranchProtectionRule.RequiredPullRequestReviews.DismissStaleReviews = &dismissStale
+		modded = true
+	}
+	codeOwner := readBoolPtr(rule.Parameters.PullRequestParameters.RequireCodeOwnerReview)
+	if codeOwner && !readBoolPtr(branchRules.RequireCodeOwnerReviews) {
+		branchRef.BranchProtectionRule.RequiredPullRequestReviews.RequireCodeOwnerReviews = &codeOwner
+		modded = true
+	}
+	lastPush := readBoolPtr(rule.Parameters.PullRequestParameters.RequireLastPushApproval)
+	if lastPush && !readBoolPtr(branchRef.BranchProtectionRule.RequireLastPushApproval) {
+		branchRef.BranchProtectionRule.RequireLastPushApproval = &lastPush
+		modded = true
+	}
+	ruleReviewers := readIntPtr(rule.Parameters.PullRequestParameters.RequiredApprovingReviewCount)
+	branchReviewers := readIntPtr(branchRules.RequiredApprovingReviewCount)
+	if ruleReviewers > branchReviewers {
+		branchRef.BranchProtectionRule.RequiredPullRequestReviews.RequiredApprovingReviewCount = &ruleReviewers
+		modded = true
+	}
+	return modded
+}
+
+func applyRequiredStatusChecksRepoRule(branchRef *clients.BranchRef, rule *repoRule) bool {
+	// If the branch already requires status checks, the rule does not matter.
+	branchRules := &branchRef.BranchProtectionRule.CheckRules
+	if readBoolPtr(branchRules.RequiresStatusChecks) {
+		return false
+	}
+
+	statusParams := rule.Parameters.StatusCheckParameters
+	if len(statusParams.RequiredStatusChecks) == 0 {
+		return false
+	}
+
+	enabled := true
+	branchRules.RequiresStatusChecks = &enabled
+	if branchRules.UpToDateBeforeMerge == nil && statusParams.StrictRequiredStatusChecksPolicy != nil {
+		copyBoolPtr(statusParams.StrictRequiredStatusChecksPolicy, &branchRules.UpToDateBeforeMerge)
+	}
+	for _, chk := range statusParams.RequiredStatusChecks {
+		if chk.Context == nil {
+			continue
+		}
+		branchRules.Contexts = append(branchRules.Contexts, *chk.Context)
+	}
+	return true
+}
+
+func ruleAdminEnforced(rule *repoRuleSet) bool {
+	if len(rule.BypassActors.Nodes) == 0 {
+		return true
+	}
+	for _, bypass := range rule.BypassActors.Nodes {
+		if readBoolPtr(bypass.OrganizationAdmin) {
+			return false
+		}
```

Comment:
Note: bypass can be `nil` here, depending on the auth token used. I was testing from another account with 
```
go run main.go --repo spencerschrock/branch-protection-repo-rule-mixed --format json --show-details --verbosity debug --checks Branch-Protection | jq
```
The repo had a repo rule which allowed deletions for a bypass list, where the bypass list was the repo admin.
```
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x8 pc=0xf72795]

goroutine 83 [running]:
github.com/ossf/scorecard/v4/clients/githubrepo.ruleAdminEnforced(...)
        ~/scorecard//clients/githubrepo/branches.go:564
github.com/ossf/scorecard/v4/clients/githubrepo.applyRepoRules(0xc00053c000, {0xc00049c060, 0x1, 0x22301a0?})
        ~/scorecard//clients/githubrepo/branches.go:500 +0x3f5
github.com/ossf/scorecard/v4/clients/githubrepo.getBranchRefFrom(0xc0005200d8, {0xc00049c060, 0x1, 0x1})
        ~/scorecard//clients/githubrepo/branches.go:425 +0x15a
github.com/ossf/scorecard/v4/clients/githubrepo.(*branchesHandler).setup.func1()
        ~/scorecard//clients/githubrepo/branches.go:300 +0x745
sync.(*Once).doSlow(0x0?, 0x7fc1199b87a8?)
        ~/sdk/go1.21.0/src/sync/once.go:74 +0xbf
sync.(*Once).Do(...)
```

When I use a PAT from my account, the same command succeeds, so there's some permission gate.

Reply:
When I use the graphQL explorer on your test repo (https://github.com/thepwagner/repo-rules-sandbox):
```
"bypassActors": {
  "nodes": [
    null
  ]
},
```

Compared to viewing something on mine (https://github.com/spencerschrock/branch-protection-repo-rule-mixed):
```
""bypassActors": {
  "nodes": [
    {
      "actor": null,
      "bypassMode": "ALWAYS",
      "organizationAdmin": false,
      "repositoryRoleName": "admin"
    }
  ]
},
```

Reply:
Interestingly, the query can still see how many bypass actors a rule has, even if it cant see who they are (if `bypass == nil`). So `ruleAdminEnforced` may be as simple as:
```
return len(rule.BypassActors.Nodes) == 0
```

Reply:
Much more simple, thanks!